src/benchmarklib/tpch/tpch_query_generator.cpp

```diff
@@ -52,6 +70,339 @@ std::string TPCHQueryGenerator::get_preparation_queries() const {
+
+      parameters.emplace_back(std::to_string(size));
+      parameters.emplace_back("'%"s + material + "'");
+      parameters.emplace_back("'"s + region + "'");
+      parameters.emplace_back("'"s + region + "'");
+      break;
+    }
+
+    case 3 - 1: {
+      const auto segment = c_mseg_set.list[segment_dist(random_engine)].text;
+      static thread_local std::uniform_int_distribution<> date_diff_dist{0, 30};
+      const auto date = calculate_date(boost::gregorian::date{1995, 03, 01}, 0, date_diff_dist(random_engine));
+
+      parameters.emplace_back("'"s + segment + "'");
+      parameters.emplace_back("'"s + date + "'");
+      parameters.emplace_back("'"s + date + "'");
+      break;
+    }
+
+    case 4 - 1: {
+      static thread_local std::uniform_int_distribution<> date_diff_dist{0, 9};
+      const auto diff = date_diff_dist(random_engine);
+      const auto begin_date = calculate_date(boost::gregorian::date{1993, 01, 01}, diff);
+      const auto end_date = calculate_date(boost::gregorian::date{1993, 01, 01}, diff + 3);
+
+      parameters.emplace_back("'"s + begin_date + "'");
+      parameters.emplace_back("'"s + end_date + "'");
+      break;
+    }
+
+    case 5 - 1: {
+      const auto region = regions.list[region_dist(random_engine)].text;
+
+      static thread_local std::uniform_int_distribution<> date_diff_dist{0, 4};
+      const auto diff = date_diff_dist(random_engine);
+      const auto begin_date = calculate_date(boost::gregorian::date{1993, 01, 01}, diff * 12);
+      const auto end_date = calculate_date(boost::gregorian::date{1993, 01, 01}, (diff + 1) * 12);
+
+      parameters.emplace_back("'"s + region + "'");
+      parameters.emplace_back("'"s + begin_date + "'");
+      parameters.emplace_back("'"s + end_date + "'");
+      break;
+    }
+
+    case 6 - 1: {
+      static thread_local std::uniform_int_distribution<> date_diff_dist{0, 4};
+      const auto diff = date_diff_dist(random_engine);
+      const auto begin_date = calculate_date(boost::gregorian::date{1993, 01, 01}, diff * 12);
+      const auto end_date = calculate_date(boost::gregorian::date{1993, 01, 01}, (diff + 1) * 12);
+
+      static std::uniform_real_distribution<> discount_dist{0.02f, 0.09f};
+      const auto discount = discount_dist(random_engine);
+
+      static thread_local std::uniform_int_distribution<> quantity_dist{24, 25};
+      const auto quantity = quantity_dist(random_engine);
+
+      parameters.emplace_back("'"s + begin_date + "'");
+      parameters.emplace_back("'"s + end_date + "'");
+      parameters.emplace_back(std::to_string(discount - 0.01f));
+      parameters.emplace_back(std::to_string(discount + 0.01f));
+      parameters.emplace_back(std::to_string(quantity));
+      break;
+    }
+
+    case 7 - 1: {
+      const auto nation1 = nations.list[nation_dist(random_engine)].text;
+      const auto nation2 = nations.list[nation_dist(random_engine)].text;
```

Comment:
The standard mandates that nation1 and nation2 should be different. this doesn't appear to ensure this.